xtensa: userspace: handle DTLB multihit and incorrect load/store ring exceptions

dcpleung opened the pull request:
This adds handlers for DTLB multihit and load/store ring exceptions. These are usually the result of having both kernel and user threads accessing the same memory page.
* In some situations, both kernel and user PTEs get pulled into the TLB caches, resulting in multihit exception. So we need to remove those entries from TLB cache and let hardware reloads the correct PTE from page table.
* For the load/store ring exception, if only the kernel PTE is populated in TLB cache, the hardware raises the load/store exception as a type of access violation. However, we need to determine manually if that is actually a valid violation by testing the associated PTE in the user page table. If it is a valid access, we need to invalidate the cached TLB entry and let the hardware reload the user PTE for access.

To test this, some context switching tests have been added to the userspace test suite.

dcpleung committed:
tests: mem_protect/userspace: add thread switching tests

This adds a few tests to test switching between threads where
they are under different memory domains. This is mainly to test
if permissions are set correctly during context switch.

By default the new tests are disabled and must be explicitly
enabled. This is due to some architectures requiring some config
changes (for example, x86 needing more reserved memory domains,
and ARM64 needing more translation tables). To avoid causing
mass CI failures, only enable platforms that have been tested,
for now.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: fatal: use named macros for PS register bits...

...instead of magic numbers when printing exception information.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: mem_protect/userspace: fix privilege stack ptr for Xtensa

The priv_stack_ptr for reading and writing tests was not set
correctly on Xtensa, and was always 0. Test would pass since
the NULL page should always generate access faults on most
Xtensa platforms. However, those tests are supposed to test
reading and writing to the privilege stack. So fix that by
populating the priv_stack_ptr correctly.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: userspace: handle load/store ring exception

When a page can be accessed by both kernel and user threads,
the autofill DTLB may contain an entry for kernel thread.
This will result in load/store ring exception when it is
accessed by user thread later. In this case, we need to
invalidate all associated TLBs related to kernel access so
hardware can reload the page table the correct permission
for user thread.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: userspace: handle DTLB multihit exception

This adds a function to handle DTLB multihit exception when
userspace is enabled, as this exception may be raised due to
the same page being able to accessed by both kernel and user
threads. The auto-refill DTLBs may contain entries for same
page, one for kernel and one for user under some situations.
We need to invalidate those existing DTLB entries so that
hardware can reload from the page table.

This is an alternative to the kconfig invalidating DTLBs on
every swap: CONFIG_XTENSA_MMU_FLUSH_AUTOREFILL_DTLBS_ON_SWAP.
Although this does not have extra processing per context
switching, exception handling itself has a high cost. So
care needs to be taken on whether to enable that kconfig.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: save EXCCAUSE/EXCVADDR in BSA during exception

This saves the EXCCAUSE and EXCVADDR into BSA during exception.
These will then be used during exception handling. The reason
for doing this instead of reading from both registers during
exception handing is that another exception will override
the value in these register (e.g. DTLB miss). When returning to
the previous exception handler, these register no longer
contains the original exception cause and address. We need to
save it so that we are actually handling the orignal exception.

Coredump also now uses the EXCCAUSE and EXCVADDR from BSA
instead of reading the registers.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: move saved FPU and HIFI registers to end of BSA

This moves the block of FPU and HIFI registers in the Base Save
Area (BSA), used during interrupts and exceptions, to the end of
the block. This is done to minimize code usage in the vector
text section. During interrupt entrance, the code stores a small
set of registers first before jumping to the long handler. When
the offset to these registers from the beginning of BSA is small
enough, the compiler will emit S32I.N instead of S32I. This
saves us one byte per store (2 vs 3 bytes). This is mainly done
to avoid overflowing the vector text area.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: userspace: save PS to thread struct later

This is simply done to conserve code space in the vector text
areas. These vector text areas are very small and we should
only put code that is absolutely necessary for interrupt and
exception entrance. The saving of PS into the thread struct
can be deferred a bit. So do that.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Compliance does not like labels in assembly.

dcpleung committed:
xtensa: mmu: fix __arch_mem_map assert message

The assert error messages when l2_page_table_map() fails are not
correct. It returns false when it cannot allocate new L2 table,
and it is not able the address having already mapped. So correct
the error message.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: mmu: cosmetic changes to page table variable names

In functions which manipulate both L1 and L2 tables, make
the variable names obvious by prefixing them with l1_ or l2_.
This is mainly done to avoid confusion when reading through
those functions.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: mmu/ptables: rename flags to attrs under arch_mem_map()

arch_mem_map() takes in some flags to describe the to-be mapped
memory regions' permissions and cache status. When the flags are
translated, they become attributes in PTEs. So for functions
being called by arch_mem_map() and beyond, rename flags to
attrs to better describe its purpose.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
xtensa: mmu: properly restore PTE attributes via reset_region()

The software bits inside PTE are used to store original PTE
attributes and ring value, and those bits are used to
restore PTE to previous state.

This modifies reset_region() to properly restore attributes
and ring value when resetting memory regions to the same as
when the system boots.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

lyakh reviewed (dismissed):
while waiting for https://github.com/thesofproject/sof/pull/10242 to finish the test, I confirm, that this PR fixes the zephyr/samples/userspace/prod_consumer test on PTL ADSP

lgirdwood reviewed (commented):
LGTM, just some minor opens.

dcpleung commented on arch/xtensa/include/xtensa_mmu_priv.h:
```diff
@@ -52,17 +52,38 @@
+ *
+ * This is relative to the SW field, not the PTE entry.
+ */
+#define XTENSA_MMU_PTE_SW_ATTR_MASK		0x0000000FU
```
This is simply the 4 bits out of the 6 bits where we store the backup attributes. The other 2 bits are used to backup the ring val.

dcpleung commented on arch/xtensa/include/xtensa_mmu_priv.h:
```diff
@@ -213,6 +213,15 @@
+ * Xtensa Instruction Set Architecture (ISA) Reference Manual
+ * 4.6.5.7 Formats for Probing MMU Option TLB Entries
+ */
+#define XTENSA_MMU_PDTLB_WAY_MASK		0xFU
```
This is not configurable and is fixed in hardware.

lgirdwood commented on arch/xtensa/core/ptables.c:
```diff
@@ -1187,6 +1187,16 @@ int arch_buffer_validate(const void *addr, size_t size, int write)
 	return mem_buffer_validate(addr, size, write, XTENSA_MMU_USER_RING);
 }
 
+void xtensa_exc_dtlb_multihit_handle(void)
+{
+	/* For some unknown reasons, using xtensa_dtlb_probe() would result in
+	 * QEMU raising privileged instruction exception. So for now, just
+	 * invalidate all auto-refilled DTLBs.
+	 */
```
I think probably the biggest (in size and frequency) user with high impact would be DMA buffers for low latency audio. i.e. DMA driver is kernel space and audio pipeline is userspace. Today the kernel owns the DMA buffer, but we could avoid expensive accesses if userspace owned the DMA buffer (as kernel will not process the audio). i.e. I am assuming we can use a userspace DMA buffer to avoid this exception path ?

lgirdwood replied:
adding @kv2019i

lyakh reviewed (dismissed):
only optional minor comments

dcpleung committed:
xtensa: rename xtensa_asm2_s.h to xtensa_asm2.inc.S

xtensa_asm2_s.h is not exactly a C header. Rename it to
xtensa_asm2.inc.S to clearly state that it is has assembly
code in it as its main purpose is to declare assembly macros.
This is being done to keep checkpatch.pl from treating it as
C file and complaining about non-C syntax.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>
